to-be-ccpp-ized/utilities/vertical_diffusion/vertical_diffusion_solver.F90

```diff
@@ -0,0 +1,136 @@
+    
+        ! Matrix to decomp from.
+        ! real(r8), intent(in) :: u(ncols,pver)
+        integer,  intent(in)    :: ncols
+        integer,  intent(in)    :: pver
+        real(r8), intent(in)    :: toSolve(ncols,pver)
+    
+        ! Coefficients for diffusion and advection.
+        !
+        ! The sizes must be consistent among all the coefficients that are
+        ! actually present, i.e. coef_q_diff and coef_q_adv should be one level
+        ! bigger than coef_q and coef_q_weight, and have the same column number.
+        real(r8), contiguous, intent(in), optional :: coef_q(:,:), &
+        coef_q_diff(:,:), coef_q_adv(:,:), coef_q_weight(:,:)
+    
+        ! Boundary conditions (optional, default to 0 flux through boundary).
+        class(BoundaryType), target, intent(in), optional :: &
+        upper_bndry, lower_bndry
+    
+        ! Objects representing boundary conditions.
+        class(BoundaryCond), intent(in), optional :: l_cond, r_cond
+    
+        real(r8) :: solution(ncols,pver)
+    
+        ! decomposition.
+        type(TriDiagDecomp) :: decomp
+    
+        ! --------------- !
+        ! Local Variables !
+        ! --------------- !
+    
+        ! Operator objects.
+        type(TriDiagOp) :: add_term
+        type(TriDiagOp) :: net_operator
+    
+        ! ----------------------- !
+        ! Main Computation Begins !
+        ! ----------------------- !
+    
+        ! A diffusion term is probably present, so start with that. Otherwise
+        ! start with an operator of all 0s.
+    
+        if (present(coef_q_diff)) then
+        net_operator = diffusion_operator(p, coef_q_diff, &
+            upper_bndry, lower_bndry)
+        else
+        net_operator = zero_operator(p%n, p%d)
+        end if
+    
+        ! Constant term (damping).
+        if (present(coef_q)) then
+        add_term = diagonal_operator(coef_q)
+        call net_operator%add(add_term)
+        end if
+    
+        ! Effective advection.
+        if (present(coef_q_adv)) then
+        add_term = advection_operator(p, coef_q_adv, &
+            upper_bndry, lower_bndry)
+        call net_operator%add(add_term)
+        end if
+    
+        ! We want I-dt*(w^-1)*A for a single time step, implicit method, where
+        ! A is the right-hand-side operator (i.e. what net_operator is now).
+        if (present(coef_q_weight)) then
+        call net_operator%lmult_as_diag(-dt/coef_q_weight)
+        else
+        call net_operator%lmult_as_diag(-dt)
+        end if
+        call net_operator%add_to_diag(1._r8)
+    
+        ! Decompose
+        decomp = TriDiagDecomp(net_operator)
+        solution = toSolve
+    
+        call net_operator%finalize()
+        call add_term%finalize()
+    
+        call decomp%left_div(solution(:ncols, :), l_cond=l_cond, r_cond=r_cond)
+        !tendency = tendency - toSolve
+    
+        ! Ensure local objects are deallocated.
+        call decomp%finalize()
```

Comment:
Perhaps the cleanup calls to `net_operator%finalize()` and `add_term%finalize()` could be moved together with this line here.